Use split glob pattern for identifier, force relative path

lukescott opened the pull request:
This PR does the following:

- Moves helper functions such as `makeImport`, `freezeNamespaceObject`, and `makeNamespaceObject` out of the main function body.

- Forces path to be relative, even if `glob:` is used.

- Removes `common-path-prefix` and `common-extname` in favor of breaking apart the glob expression. The unique portion is used.

  `GlobSync` is used instead of `glob.sync` to get more information. `glob.sync` is a helper of `GlobSync` that returns the `result.found`.

  `result.minimatch.set` is the expression split by `/`. Minimatch, glob's matching library, represents literal parts of the expression as a `string`, dynamic parts as a `RegExp`, and `**` as `GLOBSTAR` (ref constant).

  NOTE: Glob supports more than one expression. So `set` is actually a 2-D array. Since 1 expression is only used in this package, the outer array always has a length of 1. So in the code we actually use `set[0]`.

  The min and max part of the split expression is used as unique part of the path:
  - `./foo/*/bar/*/bat.js` resolves to `*/bar/*`
  - `./foo/*/bar/*.js` resolves to `*/bar/*` (if filename is included extension is removed)

lukescott commented:
Ok, so I was removing the whole extension using `nodePath.extname`. The issue with this is you could have conflicts if you had something like `foo.one.txt` and `foo.two.txt`. That was the point of `common-extname`.

The issue with `common-extname`, however, is it would remove `.one.txt` if you only had files with`.one.txt`. That would change if you added a file with `.two.txt` into the same directory. It would, at first, remove `.one.js`, then only remove `.js` because you added a file. That can be frustrating.

So instead I'm pulling the extension from the import expression. If you have `*.js` only `.js` is removed. If you have `*.one.js` then `.one.js` is removed. If you add a file with `.two.js` that file isn't included, so the behavior doesn't change.

I came across a similar problem with `common-path-prefix` with the older version. I had an include that looked like `./*/*.reducer.js`. I only had `./textField/textField.reducer.js`. The identifier was `textField`. As soon as I added `multipleChoice/multipleChoice.reducer.js` the first one became `textField$textField`.

With this PR and the extname changes the addition of files will not change the result since the determined identifier will be based on the pattern.

novemberborn commented on README.md:
```diff
@@ -44,60 +44,55 @@ have a directory layout like this:
-extension](https://github.com/novemberborn/common-extname). File-separators in
-the resulting strings are replaced by dollar signs. The directory components are
-then [converted into identifiers](https://github.com/novemberborn/identifierfy).
+Identifiers are generated for all matches using the unique portion of each file path. File extension found in the pattern is also removed.
```
"are also removed"

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+const dirname = nodePath.dirname
+const relative = nodePath.relative
+const resolve = nodePath.resolve
+const fileSeparator = nodePath.sep
```
I don't find this manual destructuring worthwhile, actually. It's not a big deal but if you'd want to change it that be great.

lukescott replied:
Restructuring is supported in Node 6. So when Node 4 is dropped it's an easier conversion:

```js
const {
  dirname,
  relative,
  resolve,
  sep as fileSeparator
} = require('path')
```

Also `nodePath` looks strange in the code to me.

novemberborn replied:
I know `nodePath` is odd… perhaps `path` in the `ImportDeclaration` visitor should be renamed `astPath` or `babelPath`.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
-const commonExtname = require('common-extname')
-const commonPathPrefix = require('common-path-prefix')
-const glob = require('glob')
+const GlobSync = require('glob/sync').GlobSync
```
`glob/sync` doesn't seem to be documented. Perhaps access this through `require('glob').GlobSync`? Though I can't see that class in the README either, perhaps use `require('glob').Glob` with the `sync` option?

lukescott replied:
The documentation also says:
> sync DEPRECATED: use glob.sync(pattern, opts) instead.

Which means if I apply the same logic to the class, I need to use the `GlobSync` class as well. It does appear that the `GlobSync` class is exported in `require('glob')` as well, so I can change that at least.

novemberborn replied:
OK, let's go with the `GlobSync` export.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
-      throw path.buildCodeFrameError('Cannot import the default member')
+function generateMembers (gm, cwd) {
+  const found = gm.found
+  const set = gm.minimatch.set[0]
```
There can be expressions that result in multiple sets:

```
❯ node -pe "require('glob').Glob('node_modules/{glob,minimatch}/package.json',{sync:true}).minimatch.set"
[ [ 'node_modules', 'glob', 'package.json' ],
  [ 'node_modules', 'minimatch', 'package.json' ] ]
```

lukescott replied:
Hmm. I should be able to get around this by changing  `makeRe`/`makeSubpathExpression` to return an array of regular expressions. I'll need to merge `extLen` into it though. The return would look like: [{regexp: {}, ext: '.js'},...]. Then we just figure out which expression matches each file.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+        let pattern = source.value
+        if (pattern.startsWith('glob:')) {
+          pattern = pattern.replace(/^glob:/, '')
+        } else if (!hasMagic(pattern)) {
```
What happens if the `glob:` starting pattern doesn't have any magic?

lukescott replied:
I could throw an error.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
-        if (node.specifiers.length === 0) {
-          path.replaceWithMultiple(members.map(member => t.importDeclaration([], t.stringLiteral(lookup[member]))))
-          return
+        if (!pattern) {
```
Given the `hasMagic()` check statement above, why is this condition reached?

lukescott replied:
This would be reached if they just had `"glob:"`. If I add a `hasMagic` when `pattern.startsWith('glob:')` is true, then I can throw an error if pattern is empty or isn't a glob pattern.

novemberborn replied:
Yea, let's clean that up above, where the pattern is being normalized.

novemberborn commented on test/import-glob.js:
```diff
@@ -89,33 +100,33 @@ test('does not transform standard import', t => {
     "import fooBar from './fixtures/foo-bar/foo-bar.txt';")
 })
 
 test('constructs the member by identifierfying the file name, including remaining extnames', t => {
```
This test seems obsolete given the change in behavior, and the new tests added at the end?

lukescott replied:
Not sure. I just maintained the existing tests. If the branch coverage stays and same and you feel they are not needed I'll remove them.

novemberborn replied:
Yea let's try it.

novemberborn commented on test/import-glob.js:
```diff
@@ -89,33 +100,33 @@ test('does not transform standard import', t => {
 
 test('constructs the member by identifierfying the file name, without the common extname', t => {
```
This test seems obsolete given the change in behavior, and the new tests added at the end?

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+        const gm = GlobSync(pattern, {cwd, strict: true})
+        const members = generateMembers(gm, cwd)
+        const unique = Object.create(null)
+        members.forEach(m => {
```
`for (const m of members) {}` instead?

lukescott replied:
Sure. This is more out of habit. When using babel the transformation of for/of adds a try/catch, which is useless for arrays. Not using babel anymore though :).

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
 
-            return makeImport(localName, lookup[importName])
+        const replacement = []
+        specifiers.forEach(s => {
```
`for/of`

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
-            makeNamespaceObject(localName, members),
-            freezeNamespaceObject(localName)
+        if (replacement.length === 0) {
+          members.forEach(
```
`for/of`

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+            // importDefaultSpecifier has previously been rejected.
+            members.forEach(
+              m => replacement.push(
+                  makeImport(t, `_${localName}_${m.name}`, m.relative)
```
```js
for (const m of members) {
  replacement.push(makeImport(t, `_${localName}_${m.name}`, m.relative))
}
```

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
-            members.map(member => makeImport(`_${localName}_${member}`, lookup[member])),
-            makeNamespaceObject(localName, members),
-            freezeNamespaceObject(localName)
+        if (replacement.length === 0) {
```
This can only be `true` if `specifiers.length === 0`. Perhaps wrap the above logic in an a `specifiers.length > 0` condition and put this in an `else` branch?

lukescott replied:
Sure. My thought on this was to only use one `path.replaceWithMultiple` at the end.

novemberborn replied:
That's cool, I was trying to highlight that checking `replacement.length` is an indirect test for `specifiers.length === 0`, so we could just have an `if`/`else` condition on that.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+  const set = gm.minimatch.set[0]
+  const regexp = new RegExp(makeRe(set), 'i')
+  const ext = extLen(set)
+  return found.map(file => {
```
Perhaps use `gm.found` directly?

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+  const regexp = new RegExp(makeRe(set), 'i')
+  const ext = extLen(set)
+  return found.map(file => {
+    let part = file.match(regexp)[1]
```
`subpath` rather than `part`?

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+  const found = gm.found
+  const set = gm.minimatch.set[0]
+  const regexp = new RegExp(makeRe(set), 'i')
+  const ext = extLen(set)
```
The naming of `ext` and `extLen` could be clearer.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+    return 0
   }
-  return pattern
+  return last._glob.match(/(?:\.[^.*?!+@[\]()]+)*$/)[0].length
```
Could you add a comment explaining what this is (not) trying to match?

novemberborn replied:
It's also unfortunate that `_glob` isn't documented in `minimatch`.

lukescott replied:
I think they provide the glob expressions on `minimatch` separately, similar to set. I'll see if I can use that instead.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+      if (p === GLOBSTAR) {
+        parts.push('(?:(?!(?:/|^)\\.).)*?')
+      } else {
+        parts.push(p._src)
```
`_src` isn't documented in `minimatch`. `p.source.slice(1, -1)` should be the same though.

lukescott replied:
Yeah, I'm assuming it only adds `^` `$`. I'll verify.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+  for (let i = 0; i < set.length; i++) {
+    const p = set[i]
+    if (typeof p !== 'string') {
+      if (p === GLOBSTAR) {
```
I would do something like:

```js
if (p === GLOBSTAR) {

} else if (typeof p === 'string') {

} else {
  // regexp
}
```

lukescott replied:
Both `GLOBSTAR` and `typeof p !== 'string'` are regular expression - GLOBSTAR is just pre-defined. I want to capture min/max on both of those. If I were to do something like above, I would be duplicating that logic.

novemberborn replied:
Fair enough. Perhaps start with `if(typeof p === 'string')` and avoid the negation. Anyway, small fry.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+const resolve = nodePath.resolve
+const fileSeparator = nodePath.sep
+
+function makeRe (set) {
```
`makeSubpathExpression`?

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+        parts.push(p._src)
+      }
+      if (min === undefined) {
+        min = i
```
Since you only assign this once you could include the `(` when first pushing the part of the expression.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+    const p = set[i]
+    if (typeof p !== 'string') {
+      if (p === GLOBSTAR) {
+        parts.push('(?:(?!(?:/|^)\\.).)*?')
```
Could you add a comment explaining this expression?

novemberborn commented on index.js:
```diff
@@ -1,125 +1,201 @@
+      }
+      max = i
+    } else {
+      parts.push(p.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&'))
```
What does this replacement achieve?

lukescott replied:
This escapes the literal string. I had this wrapped in a function, but I only used it once. I could make it a function again. Or comment it. I pulled it from minimatch.

novemberborn replied:
A function would be good, that tends to make it self-documenting. Would be good to reference a specific line in the minimatch source as well.

lukescott commented:
@novemberborn I'm done with the first pass. I mostly focused on adding brace support. Several of the items above are fixed, although with the code changes I'm sure there will be more comments.

I did not remove the two tests yet - I think it may be a good idea to go thru each test and clean it up a bit. I put in a couple brace tests using existing fixtures, which may not make much sense. When I have more time I can go over the tests in more detail.

novemberborn commented on index.js:
```diff
@@ -1,125 +1,230 @@
+      return true
     }
   }
+  return false
```
I suppose this could be `return specifiers.some(s => s.type === 'ImportDefaultSpecifier')`.

novemberborn commented on test/import-glob.js:
```diff
@@ -153,7 +164,67 @@ Object.freeze(members);`)
+  t.is(
+    transform("import * as members from 'glob:./fixtures/ext-from-pattern/*.{txt,csv}'"),
+    `import _members_oneFoo from './fixtures/ext-from-pattern/one.foo.txt';
+import _members_twoFoo from './fixtures/ext-from-pattern/two.foo.txt';
```
There should be a `three.csv` file here.

lukescott commented:
I made all the review changes mentioned above. At some point I want to revisit the tests and go over them in more detail, but that shouldn't hold up the merge.

Sorry for the delay. I've been dealing with a family crisis since last week.

novemberborn commented:
Thanks, I'll try and have a look tomorrow.

> Sorry for the delay. I've been dealing with a family crisis since last week.

No worries, that's far more important. Hope it ends well.

novemberborn commented:
I've pushed some tweaks, mostly renaming the very short variables which, for me at least, helps with readability.

I've also tried to refactor the `makeSubpathExpressions` function but I'm a bit confused by parts of it. I've added `FIXME` comments for that. Perhaps you could elaborate on those?

I'd like to take a pass through the README to make sure it's phrased in terms of the new glob-determines-name behavior. I haven't looked in depth at the tests yet.

lukescott commented on index.js:
```diff
@@ -1,125 +1,211 @@
+        // other condition?
+        // FIXME: What is the logic here? {a,b} will lead to capture, as will
+        // {a,a*}. Is this to do with multiple braces being expanded? E.g.
+        // {a,b}/foo/{c,d}?
```
If you have `/*/bar.{txt,csv}` then `bar.txt` and `bar.csv` are strings. So they don't get captured, unless we compare between expressions to see that they are different.

lukescott commented on index.js:
```diff
@@ -1,125 +1,211 @@
+
+    const last = expressions.length - 1
+    // FIXME: Why check for strings? What is the significance of the expressions[last] !== nextExpressions[last]?
+    // Why not use path.extname()?
```
Same reasoning as above. The filenames are braced, so they are captured. So we want to get the extension. We only grab the extension if the filename is captured.

lukescott commented on index.js:
```diff
@@ -1,125 +1,211 @@
+      : null
+
+    return {
+      // FIXME: Why even return extname? Can't it be excluded from the regexp match?
```
Hmmmm.... I mean we could by using `minimatch.set` instead. I'm just concerned the regexp could change. `minimatch.globParts` is just the brace-expanded glob pattern, which should stay consistent. If we were to use `minimatch.set`, we would have to de-duplicate the expressions.

lukescott commented on index.js:
```diff
@@ -1,125 +1,211 @@
-          path.replaceWithMultiple(members.map(member => t.importDeclaration([], t.stringLiteral(lookup[member]))))
-          return
+        if (pattern.startsWith('glob:')) {
+          pattern = pattern.replace(/^glob:/, '')
```
This could possibly be:
```js
pattern = pattern.substr('glob:'.length)
```
`glob:` could be changed to a constant. I think the regexp might be a tad overkill here.

lukescott commented:
> I've also tried to refactor the makeSubpathExpressions function but I'm a bit confused by parts of it. I've added FIXME comments for that. Perhaps you could elaborate on those?

My earlier comments no longer show. Basically, yes, that code was for matching braced patterns. The logic was basically if one of the other patterns did not match, it was dynamic even if it was a string. Which was only a problem if it was at the end. For exampe,`foo.{csv,txt}` produces `foo.csv` and `foo.txt`.

I refactored the code to stop using `globSet` (not documented anyway) and I match the extension from the RegExp pattern instead. I also place the ending `)` before the extension, so I no longer have to treat extension separately. Also it now creates a single RegExp for all sets.

novemberborn commented:
I've been at a conference this weekend, will try to have a look at this tomorrow.

Meanwhile, is it OK if I add you as a maintainer on this repo and npm? These are really great contributions.

novemberborn commented:
@lukescott very nice!

I've pushed a fix for patterns like `{a,b/c}/d`, in which case the `minimatch.set` array contains pattern lists of varying lengths. This was quite tricky to work through so I ended up rewriting a fair bit of your recent code. I'm also using `Set()` to obviate the need for `lodash.uniq`. Let me know what you think.

novemberborn commented:
(Still need to review tests and update the readme).

lukescott commented:
Hmmm. I didn't consider `/` inside the brace. What happens with multiple instances of those in different positions? minimatch's makeRe takes separate patterns and combines them with `|`. With flattenSet I was trying to make a shorter pattern. I think it may be easier (and more readable) to revert it to the original `makeSubpathExpressions` and simply join each expression with a `|`.

lukescott commented:
Hmmm... I didn't consider `/` in path. What happens when you have more than one of these? I'm not sure I fully understand what's going on in the revised code.

lukescott commented:
Here are a couple interesting patterns:
```
./fixtures/{use-dir-name,}/{one,two,thr/ee}/foo.txt
./fixtures/{use-dir-name,use-dir/name2}/{one,two,thr/ee}/foo.txt
```
Which generates
```
^(?:\.)/(?:fixtures)/((?:use-dir-name|one|two|thr)(?:/(?:one|two|thr|ee))?(?:/(?:ee))?)/(?:foo\.txt)$
^(?:\.)/(?:fixtures)/((?:use-dir-name|use-dir)/(?:one|two|thr|name2)(?:/(?:ee|one|two|thr))?(?:/(?:ee))?)/(?:foo\.txt)$
```
The expressions are less precise. They allow for invalid matches... although glob has already done matching, so maybe it's ok?

minimatch's makeRe makes each expression separately and joins with `|`. It's a longer expression though:

```
/^(?:\.\/fixtures\/use\-dir\-name\/one\/foo\.txt|\.\/fixtures\/use\-dir\-name\/two\/foo\.txt|\.\/fixtures\/use\-dir\-name\/thr\/ee\/foo\.txt|\.\/fixtures\/use\-dir\/name2\/one\/foo\.txt|\.\/fixtures\/use\-dir\/name2\/two\/foo\.txt|\.\/fixtures\/use\-dir\/name2\/thr\/ee\/foo\.txt)$/
/^(?:\.\/fixtures\/use\-dir\-name\/one\/foo\.txt|\.\/fixtures\/use\-dir\-name\/two\/foo\.txt|\.\/fixtures\/use\-dir\-name\/thr\/ee\/foo\.txt|\.\/fixtures\/one\/foo\.txt|\.\/fixtures\/two\/foo\.txt|\.\/fixtures\/thr\/ee\/foo\.txt)$/
```

The reason we don't use minimatch's makeRe is because it lacks capture / extension isolation.

Perhaps we can find a solution in the middle. Identify the captured portion and expand only that part of the expression. So the above would look like this:

```
^\./fixtures/(use-dir-name/one|use-dir-name/two|use-dir-name/thr/ee|one|two|thr/ee)/foo\.txt$
^\./fixtures/(use-dir-name/one|use-dir-name/two|use-dir-name/thr/ee|use-dir/name2/one|use-dir/name2/two|use-dir/name2/thr/ee)/foo\.txt$
```

It might actually be a good idea to split the expression generator off into it's own thing.

lukescott commented:
I created a package called `minimatch-capture`, which now does most of the heavy lifting for capturing. This package has been slimmed down quite a bit. `minimatch-capture` has 67 tests covering varying use cases.

One change is braces are no longer considered capturable by themselves. For instance, with the path `./fixtures/use-dir-name/{one,two,thr/ee}/foo.txt` there is nothing to capture because it is split into 3 patterns.

In that test I changed it to `./fixtures/use-dir-name/{!(sub),@(sub)/*}/foo.txt`, which captures:
```
./fixtures/use-dir-name/one/foo.txt = one
./fixtures/use-dir-name/two/foo.txt = two
./fixtures/use-dir-name/sub/three/foo.txt = sub/three
./fixtures/use-dir-name/sub/four/foo.txt = sub/four
```
You can change the pattern to `./fixtures/use-dir-name/{!(sub),sub/*}/foo.txt`, which captures:
```
./fixtures/use-dir-name/one/foo.txt = one
./fixtures/use-dir-name/two/foo.txt = two
./fixtures/use-dir-name/sub/three/foo.txt = three
./fixtures/use-dir-name/sub/four/foo.txt = four
```

The reasons for this:

1. The test `./fixtures/ext-from-pattern/*.{txt,csv}` wants to omit the extension. `minimatch-capture` doesn't care about extension.

2. When I tried to make `minimatch-capture`'s `split` method include braces and figure out the left/middle/right before expanding the pattern, it broke a lot of tests. The matching was no longer consistent with minimatch.

novemberborn commented:
Hey @lukescott, apologies for dropping off the radar. Will try to get back to you on this tomorrow.

novemberborn commented:
> I created a package called minimatch-capture, which now does most of the heavy lifting for capturing. This package has been slimmed down quite a bit. minimatch-capture has 67 tests covering varying use cases.

Awesome!

> One change is braces are no longer considered capturable by themselves. For instance, with the path `./fixtures/use-dir-name/{one,two,thr/ee}/foo.txt` there is nothing to capture because it is split into 3 patterns.
> 
> In that test I changed it to `./fixtures/use-dir-name/{!(sub),@(sub)/*}/foo.txt`

If I'm understanding this correctly, `*` and `**` are automatically captured. You can use `!(pattern)`, `?(pattern)`, `+(pattern)`, `*(pattern)` and `@(pattern)` to ensure other path components are captured. Brace expansions are not necessarily captured, unless they're between other capture groups. For example `./*/{foo,bar}/*.txt` causes `foo` and `bar` to be included in the identifiers. Otherwise it's not possible to capture brace expansions (`({foo,bar})` doesn't work), so they have to be rewritten using the parentheses-matching syntax.

I've updated the tests and the readme, though we should probably document the capturing behavior better, especially as it comes to brace expansions. Beyond that I think this is good to go!

lukescott commented:
> If I'm understanding this correctly, * and ** are automatically captured. You can use !(pattern), ?(pattern), +(pattern), *(pattern) and @(pattern) to ensure other path components are captured. Brace expansions are not necessarily captured, unless they're between other capture groups. For example ./*/{foo,bar}/*.txt causes foo and bar to be included in the identifiers. Otherwise it's not possible to capture brace expansions (({foo,bar}) doesn't work), so they have to be rewritten using the parentheses-matching syntax.

Yep, that's correct.

novemberborn commented:
Landed! 🚀 ✨ 🎉  And shipped as 2.0.0!

I've added you to this repo and on npm as well. Many many thanks for pushing this new approach.
